### PR TITLE
add Contacts SwiftPM package

### PR DESCRIPTION
**Pull request by JuliaSamartseva**
- ContactList class was moved to package
- ContactList class was made persistent
- Added logging using SwiftLog library
- All tests pass

### CODE REVIEW

File: juliasamartseva/Contacts/Package.swift [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  name: "Contacts",
+  products: [
+    // Products define the executables and libraries produced by a package, and make them visible to other packages.
+    .library(
+      name: "Contacts",
+      targets: ["Contacts"])
+  ],
+  dependencies: [
+    // Dependencies declare other packages that this package depends on.
+    // .package(url: /* package url */, from: "1.0.0"),
```
Comment (gartemiy) [inline juliasamartseva/Contacts/Package.swift:16]: NIt: you may remove the comments as it is clear you know how to use it) Here and in other places as well.

File: juliasamartseva/Contacts/Sources/Contacts/Contacts.swift [hunk 2]
```diff
@@ -52,27 +55,36 @@ public struct WorkInfo: Codable, Equatable {
...
     return contacts.count
   }
-
-  init(contacts: Contact...) {
-    self.contacts = contacts
-  }
-
-  init(contacts: [Contact]) {
-    self.contacts = contacts
+  let logger = Logger(label: "com.google.Contacts.ContactList")
```
Comment (gartemiy) [inline juliasamartseva/Contacts/Sources/Contacts/Contacts.swift:63]: don't you need it as private?

File: juliasamartseva/Contacts/Sources/Contacts/Contacts.swift [hunk 3]
```diff
@@ -52,27 +55,36 @@ public struct WorkInfo: Codable, Equatable {
...
+
+  public init() throws {
+    logger.info("Creating contact from file")
+    fileURL = try Self.makeDefaultURL()
+    do {
+      contacts = try [Contact](jsonFileURL: fileURL)
+    } catch CocoaError.fileReadNoSuchFile {
+      contacts = []
+    } catch let error {
+      logger.error("Unidentified exception was thrown: \(error)")
```
Comment (gartemiy) [inline juliasamartseva/Contacts/Sources/Contacts/Contacts.swift:73]: Maybe rework the message a little bit to state what phase the library was in when error happened. For example "Failed to initialize the contacts with: \(error)"

File: juliasamartseva/Contacts/Sources/Contacts/Contacts.swift [hunk 4]
```diff
@@ -1,10 +1,11 @@
 import Foundation
+import Logging
 
 // MARK: - Contact class
 
 public struct Contact: Codable, Equatable {
 
-  public let id = UUID()
+  public let id: UUID
```
Comment (gartemiy) [inline juliasamartseva/Contacts/Sources/Contacts/Contacts.swift:8]: Do you need this change? Inlined version looks good

File: juliasamartseva/Contacts/Sources/Contacts/Contacts.swift [hunk 5]
```diff
@@ -52,27 +55,36 @@ public struct WorkInfo: Codable, Equatable {
 // MARK: - ContactList class
 
 public class ContactList {
-
+  let fileURL: URL
```
Comment (gartemiy) [inline juliasamartseva/Contacts/Sources/Contacts/Contacts.swift:58]: You may keep it private as not used in the tests.

File: juliasamartseva/Contacts/Sources/Contacts/Contacts.swift [hunk 6]
```diff
@@ -52,27 +55,36 @@ public struct WorkInfo: Codable, Equatable {
...
   public internal(set) var contacts = [Contact]()
   var size: Int {
     return contacts.count
   }
```
Comment (gartemiy) [inline juliasamartseva/Contacts/Sources/Contacts/Contacts.swift:62]: This should be `public` as well, shouldn't it?

File: juliasamartseva/Contacts/Sources/Contacts/Contacts.swift [hunk 7]
```diff
@@ -52,27 +55,36 @@ public struct WorkInfo: Codable, Equatable {
...
-    self.contacts = contacts
+  let logger = Logger(label: "com.google.Contacts.ContactList")
+
+  public init() throws {
+    logger.info("Creating contact from file")
+    fileURL = try Self.makeDefaultURL()
+    do {
+      contacts = try [Contact](jsonFileURL: fileURL)
+    } catch CocoaError.fileReadNoSuchFile {
+      contacts = []
```
Comment (gartemiy) [inline juliasamartseva/Contacts/Sources/Contacts/Contacts.swift:71]: Super nit: maybe log that new list is initialised.

File: juliasamartseva/Contacts/Sources/Contacts/Contacts.swift [hunk 8]
```diff
@@ -81,78 +93,71 @@ public class ContactList {
...
     if let index = contacts.firstIndex(where: { $0.id == id }) {
       contacts[index].name = newContact.name
       contacts[index].phone = newContact.phone
       contacts[index].email = newContact.email
       contacts[index].work = newContact.work
+      try writeToFile()
+      logger.info("Successfully changed contact with id \(id) to new contact")
+    } else {
+      logger.info("No contact with \(id) was found")
     }
```
Comment (gartemiy) [inline juliasamartseva/Contacts/Sources/Contacts/Contacts.swift:110]: Back to ids and your example. If you just want to change the name of the contact, like in your tests
```
    guard
      let sergii = try? Contact(name: "Sergii", phone: "380660989878", email: "df@g.com", work: nil)
    else {
      XCTFail("The contacts initialization throws error")
      return
    }
    try contactList.add(sergii)
    let sergiiId = sergii.id

    // WHEN
    let newContact = try Contact(
      name: "Sergey", phone: "380660989878", email: "df@g.com", work: nil)
    try contactList.changeContact(by: sergiiId, to: newContact)
```
Why should client bother with ids?

Much simpler API would be to update the contacts object and to update it without any id:
```
    guard
      var sergii = try? Contact(name: "Sergii", phone: "380660989878", email: "df@g.com", work: nil)
    else {
      XCTFail("The contacts initialization throws error")
      return
    }
    try contactList.add(sergii)

    // WHEN
    sergii.name = "Sergey"
    try contactList.updateContact(sergii)
```

File: juliasamartseva/Contacts/Tests/ContactsTests/XCTestManifests.swift [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import XCTest
+
+#if !canImport(ObjectiveC)
+  public func allTests() -> [XCTestCaseEntry] {
+    return [
+      testCase(ContactsTests.allTests)
+    ]
+  }
+#endif
```
Comment (gartemiy) [inline juliasamartseva/Contacts/Tests/ContactsTests/XCTestManifests.swift:9]: Don't you need to remove it as you removed the property `allTests` and in `LinuxMain.swift` you may add `fatalError` to indicate that these are not supported yet.

File: juliasamartseva/Contacts/Sources/Contacts/Contacts.swift (hunk 4 above)
Comment (JuliaSamartseva) [reply juliasamartseva/Contacts/Sources/Contacts/Contacts.swift:8]: After reading contacts from file, the program would create a new id if I didn't make this change. However, we need the id's after reading from file to stay the same as we have written them.

**Pull request merged**